Fixed #12810 -- Added a check for clashing ManyToManyField.db_table names.

berkerpeksag opened the pull request:
https://code.djangoproject.com/ticket/12810

charettes commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,25 @@ class ParkingLot(Place):
+        expected = [
+            Error(
+                "Field is using a table that has already been registered by "
+                "<class 'invalid_models_tests.test_models.OtherModelTests.test_m2m_table_name_clash.<locals>.Bar'>.",
```
This assertion will fail on Python 2 as the `repr`esentation of locally defined classes is not the same.

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,22 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            for model in self.opts.apps.get_models(include_auto_created=True)
+        }
+        if self.db_table in set(registered_tables.keys()):
+            msg = 'Field is using a table that has already been registered by {!r}.'
```
I would use the model `_meta.app_label` and `object_name` instead of its representation.

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,22 @@ def _check_relationship_model(self, from_model=None, **kwargs):
         return errors
 
+    def _check_table_uniqueness(self, **kwargs):
+        registered_tables = {
```
Abort early if `self.db_table is None`

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,22 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+    def _check_table_uniqueness(self, **kwargs):
+        registered_tables = {
+            model._meta.db_table: model
+            for model in self.opts.apps.get_models(include_auto_created=True)
```
Shouldn't you exclude `self.rel.through` from here? Else your automatically created table will always be present in `registered_tables`.

berkerpeksag replied:
I went with `self.remote_field.through` instead.

charettes commented:
It would be great to see a test with two models defining `ManyToManyField` with a clashing `db_table` to test that `include_auto_created=True` is passed to `get_models()`.

berkerpeksag commented:
Thanks for the review. I've now addressed all of your review comments.

> It would be great to see a test with two models defining ManyToManyField with a clashing db_table to test that include_auto_created=True is passed to get_models().

Wouldn't such a test case be detected by fields.E303?

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,26 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            if model != self.remote_field.through
+        }
+        if self.db_table in set(registered_tables.keys()):
+            msg = 'Field is using a table that has already been registered by \'{}.{}\'.'
```
I would stick to `%` formating for the sake of consistency with other `_check` methods.

charettes commented:
> Wouldn't such a test case be detected by fields.E303?

Here's what I had in mind

```python
class Foo(models.Model):
    pass

class Bar(models.Model):
    foos = models.ManyToManyField(Foo, db_table='clash')

class Baz(models.Model):
    foos = models.ManyToManyField(Foo, db_table='clash')
```

charettes commented:
It would be great if the message pointed to the m2m field with the `db_table` clash instead of the auto-created intermediary model in this case. e.g. `Field is using a table that has already been registered by app.Baz.foos`

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,26 @@ def _check_relationship_model(self, from_model=None, **kwargs):
 
+    def _check_table_uniqueness(self, **kwargs):
+        if self.db_table is None:
+            return []
```
From a second read on the ticket it looks like we should always validate the table name even when not explicitly specified. You can remove this early return.

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,26 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            for model in self.opts.apps.get_models(include_auto_created=True)
+            if model != self.remote_field.through
+        }
+        if self.db_table in set(registered_tables.keys()):
```
You should use `self.m2m_db_table` instead of `self.db_table`.

charettes commented:
Another test for the initially reported case would be required. That is when the auto-generated table name clashes

```python
class Foo(models.Model):
    class Meta:
        db_table = 'bar_foos'

class Bar(models.Model):
    # In this case the auto-generated `db_table` will be bar_foos
    foos = models.ManyToManyField(Foo)

    class Meta:
        db_table = 'bar'
```

berkerpeksag commented:
> Another test for the initially reported case would be required. That is when the auto-generated table name clashes

I think this is basically the same same as https://code.djangoproject.com/ticket/20098 right?

berkerpeksag commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,24 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            for model in self.opts.apps.get_models(include_auto_created=True)
+            if model != self.remote_field.through
+        }
+        if not isinstance(self.remote_field.through, str) and self.m2m_db_table() in set(registered_tables.keys()):
```
I added the isinstance check because test_missing_relationship_model in invalid_models_tests.test_relative_fields.RelativeFieldTests was failing with the following traceback:

```
Traceback (most recent call last):
  File "/Users/berkerpeksag/projects/django/tests/invalid_models_tests/test_relative_fields.py", line 299, in test_missing_relationship_model
    errors = field.check(from_model=Group)
  File "/Users/berkerpeksag/projects/django/django/db/models/fields/related.py", line 1143, in check
    errors.extend(self._check_table_uniqueness(**kwargs))
  File "/Users/berkerpeksag/projects/django/django/db/models/fields/related.py", line 1385, in _check_table_uniqueness
    print('m2m_db_table', self.m2m_db_table())
  File "/Users/berkerpeksag/projects/django/django/utils/functional.py", line 15, in _curried
    return _curried_func(*(args + moreargs), **dict(kwargs, **morekwargs))
  File "/Users/berkerpeksag/projects/django/django/db/models/fields/related.py", line 1476, in _get_m2m_db_table
    return self.remote_field.through._meta.db_table
AttributeError: 'str' object has no attribute '_meta'
```

Please let me know if there is a better way.

charettes replied:
I suggest you skip the check (`return []` early) if the intermediary model (`self.remote_field.through`) is not resolved.

That is `isinstance(self.remote_field.through, six.string_types)`. Also I would store `m2m_db_table` in a variable as you'll need to reuse it to lookup `registered_tables` below.

berkerpeksag replied:
> Also I would store m2m_db_table in a variable as you'll need to reuse it to lookup registered_tables below.

Oh, thanks, I forgot to update that line.

charettes commented:
> I think this is basically the same same as https://code.djangoproject.com/ticket/20098 right?

Both tickets are related but are dealing with different issues. [#20098](https://code.djangoproject.com/ticket/20098) is about model table clash. It's another can of worm as the error will need to be reported at the model level in this case and will need to make sure to special case unmanaged models (`_meta.managed = False`).

Auto-created intermediary table names are an implementation detail of `ManyToManyField` and should be reported as `fields.E340` errors.

berkerpeksag commented:
Thanks for the clarification, Simon. I think I've now addressed all of your comments.

charettes commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,69 @@ class ParkingLot(Place):
+        expected = [
+            Error(
+                "Field is using a table that has already been registered by "
+                "'invalid_models_tests.Bar.bar'.",
```
`Bar.bar` isn't a field. The message should be `Field is using a table that has already been registered by 'invalid_models_tests.Bar'` as the `db_table` is declared on a non-`auto_created` model (`_meta.auto_created is False`).

charettes commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,69 @@ class ParkingLot(Place):
+        expected = [
+            Error(
+                "Field is using a table that has already been registered by "
+                "'invalid_models_tests.Baz_foos.foos'.",
```
As the `db_table` is attached to an `auto_created` model the message should be `Field is using a table that has already been registered by 'invalid_models_tests.Baz.foos'.`. The intermediary model `Baz_foos` is an implementation detail and shouldn't be referred to, plus it has no `foos` field.

charettes commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,69 @@ class ParkingLot(Place):
+                id='fields.E340',
+            )
+        ]
+        self.assertEqual(errors, expected)
```
I would also check `Baz.check()` errors while you're at it.

berkerpeksag replied:
Do you mean Foo.check() or the Baz model in test_m2m_field_table_name_clash? I went with the latter but let me know if I got this wrong.

charettes replied:
I meant exactly what you did, I'm not sure how my comment got attached here.

charettes commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,69 @@ class ParkingLot(Place):
+        expected = [
+            Error(
+                "Field is using a table that has already been registered by "
+                "'invalid_models_tests.Foo.foos'.",
```
The `Foo` model as no `foos` field. As `Foo` is not `auto_created` it should have a similar message to `test_m2m_table_name_clash`.

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,27 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+        m2m_db_table = self.m2m_db_table()
+        if m2m_db_table in set(registered_tables.keys()):
+            msg = 'Field is using a table that has already been registered by \'%s.%s.%s\'.'
+            table = registered_tables[m2m_db_table]
```
The `table` variable is actually a `models.Model` instance so it might be good to rename it to `model`.

In the case of auto-created models `model._meta.auto_created` will be pointing at the model at the origin of the creation else it will be `False`. When it's `False` the resulting message should be of the form `(opts.app_label, opts.object_name)` else it should be of the form `(opts.app_label, opts.object_name, field.name)` where `field` is retrieved from iterating over `model._meta.auto_created._meta.many_to_many` where `field.remote_field.through is model`.

charettes commented on docs/ref/checks.txt:
```diff
@@ -239,6 +239,8 @@ Related Fields
   ``<field name>``.
 * **fields.E339**: ``<model>.<field name>`` is not a foreign key to ``<model>``.
+* **fields.E340**: Field is using a table that has already been registered
+  by ``<model>``.
```
I'm not sure how to express it but the message should mention that it can be formulated as `<model>.<field>` when the conflicting model is `auto_created`.

charettes replied:
Looking back at other documented I think "Field is using a table that has already been registered by ``<model>/<model>.<field name>``"

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,27 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            table = registered_tables[m2m_db_table]
+            return [
+                checks.Error(
+                    msg % (table._meta.app_label, table._meta.object_name, self.name),
```
You can replace `table._meta.app_label` and `table._meta.object_name` by `table._meta.label`

charettes commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,37 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+            if model != self.remote_field.through
+        }
+        m2m_db_table = self.m2m_db_table()
+        if m2m_db_table in set(registered_tables.keys()):
```
No need to create a `set()` here.

berkerpeksag replied:
`.keys()` can also be removed. Do you want to keep it?

charettes replied:
`if m2m_db_table in registered_tables` should do!

berkerpeksag replied:
Done.

charettes commented:
LGTM

timgraham commented on docs/ref/checks.txt:
```diff
@@ -239,6 +239,8 @@ Related Fields
 * **fields.E338**: The intermediary model ``<through model>`` has no field
   ``<field name>``.
 * **fields.E339**: ``<model>.<field name>`` is not a foreign key to ``<model>``.
+* **fields.E340**: The field's intermediary table name clashes with
```
Not sure which format you settled on, but this is different from the actual error message. It might be helpful to include the table name in the message.

berkerpeksag replied:
They are both correct but the formatting can be improved. How about this?

```diff
diff --git a/docs/ref/checks.txt b/docs/ref/checks.txt
index bbbb6d8..6cd46ce 100644
--- a/docs/ref/checks.txt
+++ b/docs/ref/checks.txt
@@ -240,7 +240,7 @@ Related Fields
   ``<field name>``.
 * **fields.E339**: ``<model>.<field name>`` is not a foreign key to ``<model>``.
 * **fields.E340**: The field's intermediary table name clashes with
-  ``<model>/<model>.<field name>`` table name.
+  ``<model>`` or ``<model>.<field name>`` table name.
 * **fields.W340**: ``null`` has no effect on ``ManyToManyField``.
 * **fields.W341**: ``ManyToManyField`` does not support ``validators``.
 * **fields.W342**: Setting ``unique=True`` on a ``ForeignKey`` has the same
```

timgraham replied:
I was referring to the difference between "Field is using a table that has already been registered by" (code) and "The field's intermediary table name clashes with" (docs).

berkerpeksag replied:
I personally prefer the one in checks.txt. @charettes which one do you prefer?

charettes replied:
I also prefer the checks.txt version but as Tim said we should include the table name in the message. What about "The field's intermediary table ``<table name>`` clashes with ``<model>`` or ``<model>.<field name>`` table name."?

timgraham replied:
Looks good, just "'s" in "with <thing>'s table name."

berkerpeksag replied:
Done, thanks.

timgraham commented on tests/invalid_models_tests/test_models.py:
```diff
@@ -761,3 +761,75 @@ class ParkingLot(Place):
+                id='fields.E340',
+            )
+        ]
+        self.assertEqual(errors, expected)
```
I'd use:
```
self.assertEqual(Bar.check(), expected)
```
to save a line.

berkerpeksag replied:
Agreed but I was following the current style in the file.

berkerpeksag replied:
Do you still want me to remove `errors = ...` from test_m2m_table_name_clash and test_m2m_autogenerated_table_name_clash or are they OK to be compatible with the rest of the file?

timgraham replied:
I'd use the preferred style so at least future tests might copy it.

berkerpeksag replied:
I updated tests to use `self.assertEqual(Model.check(), expected)` in my branch. I will push once we resolve the exception message difference in checks.txt.

timgraham commented on django/db/models/fields/related.py:
```diff
@@ -1375,6 +1376,37 @@ def _check_relationship_model(self, from_model=None, **kwargs):
+                        if field.remote_field.through is model:
+                            return field.name
+                opts = model._meta.auto_created._meta
+                msg_suffix = '\'%s.%s\'.' % (
```
Use outer double quotes to avoid backslashes.

timgraham replied:
msg_suffix -> clashing_obj (or similar?)

berkerpeksag replied:
Done, thanks.

timgraham commented:
merged in 0bce2f102c5734d0c7ff1ebf0b10a316d58ed5ce, thanks!
